streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -130,7 +135,11 @@ private boolean productionExceptionIsFatal(final Exception exception) {
         String errorMessage = EXCEPTION_MESSAGE;
         // There is no documented API for detecting retriable errors, so we rely on `RetriableException`
         // even though it's an implementation detail (i.e. we do the best we can given what's available)
-        if (exception instanceof RetriableException) {
+        if (exception instanceof TimeoutException) {
+            final String topicTimeoutHint = String.format(TIMEOUT_HINT_TEMPLATE, topic);
+            errorLogMessage += topicTimeoutHint;
+            errorMessage += topicTimeoutHint;
+        } else if (exception instanceof RetriableException) {
```

Comment:
I moved the comment below as you asked. However, on a personal note I think the both blocks are the right ones  as TimeoutException is a subclass of Retriable Exception.

Reply:
I see your point, but it seems like the comment is more acknowledging that using `RetriableException` as a catch-all for all "retriable exceptions" isn't guaranteed to work, although it's the best we can do right now. If I understood the consumer API, though, `TimeoutException` _is_ recommended as the way to catch "timeout exceptions".

But this might also be me justifying the position after the fact... Anyway, thanks for making the change.